plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1359 @@
+        """
+        The electron temperature over the course of the simulation in
+        kelvin.
+
+        The index of this array corresponds to the time step.
+        """
+        return self._T_e
+
+    @property
+    def time(self) -> u.Quantity:
+        """
+        The time for each time step over the course of the simulation
+        in units of seconds.
+        """
+        return self._time
+
+
+class NEI:
+    r"""
+    Perform and analyze a non-equilibrium ionization simulation.
+
+    Parameters
+    ----------
+    inputs
+
+    T_e: astropy.units.Quantity or callable
+        The electron temperature, which may be a constant, an array of
+        temperatures corresponding to the times in `time_input`, or a
+        function that yields the temperature as a function of time.
+
+    n: astropy.units.Quantity or callable
+        The number density multiplicative factor.  The number density of
+        each element will be ``n`` times the abundance given in
+        ``abundances``.  For example, if ``abundance['H'] = 1``, then this
+        will correspond to the number density of hydrogen (including
+        neutral hydrogen and protons).  This factor may be a constant,
+        an array of number densities over time, or a function that
+        yields a number density as a function of time.
+
+    time_input: astropy.units.Quantity, optional
+        An array containing the times associated with ``n`` and ``T_e`` in
+        units of time.
+
+    time_start: astropy.units.Quantity, optional
+        The start time for the simulation.  If density and/or
+        temperature are given by arrays, then this argument must be
+        greater than ``time_input[0]``.  If this argument is not supplied,
+        then ``time_start`` defaults to ``time_input[0]`` (if given) and
+        zero seconds otherwise.
+
+    time_max: astropy.units.Quantity
+        The maximum time for the simulation.  If density and/or
+        temperature are given by arrays, then this argument must be less
+        than ``time_input[-1]``.
+
+    max_steps: `int`
+        The maximum number of time steps to be taken during a
+        simulation.
+
+    dt: astropy.units.Quantity
+        The time step.  If ``adapt_dt`` is `False`, then ``dt`` is the
+        time step for the whole simulation.
+
+    dt_max: astropy.units.Quantity
+        The maximum time step to be used with an adaptive time step.
+
+    dt_min: astropy.units.Quantity
+        The minimum time step to be used with an adaptive time step.
+
+    adapt_dt: `bool`
+        If `True`, change the time step based on the characteristic
+        ionization and recombination time scales and change in
+        temperature.  Not yet implemented.
+
+    safety_factor: `float` or `int`
+        A multiplicative factor to multiply by the time step when
+        ``adapt_dt`` is `True`.  Lower values improve accuracy, whereas
+        higher values reduce computational time.  Not yet implemented.
```

Comment:
Would it make sense to consider an existing ODE solver? https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html ?

Reply:
The eigenvalue method that is implemented here has the really nice mathematical property that if you take an extremely long time step, then you will reach the equilibrium ionization fractions, which means it's unconditionally stable.  It's something that we can do that's special to the set of governing equations.  With that said, it would be useful to have the option to use existing ODE solvers from SciPy, including for testing purposes.  

 - [ ] Create issue on implementing alternative ODE solvers using SciPy
 - [ ] Create issue to add narrative documentation on eigenvalue method